File: code/go/ch09/src/luago/state/closure.go
package state

import . "luago/api"
import "luago/binchunk"

type closure struct {
	proto  *binchunk.Prototype // lua closure
	goFunc GoFunction          // go closure
}

func newLuaClosure(proto *binchunk.Prototype) *closure {
	return &closure{proto: proto}
}

func newGoClosure(f GoFunction) *closure {
	return &closure{goFunc: f}
}

File: code/java/ch10/src/main/java/com/github/zxh0/luago/state/Closure.java
package com.github.zxh0.luago.state;

import com.github.zxh0.luago.api.JavaFunction;
import com.github.zxh0.luago.binchunk.Prototype;
import lombok.Getter;

@Getter
class Closure {

    final Prototype proto;
    final JavaFunction javaFunc;

    // Lua Closure
    Closure(Prototype proto) {
        this.proto = proto;
        this.javaFunc = null;
    }

    Closure(JavaFunction javaFunc) {
        this.proto = null;
        this.javaFunc = javaFunc;
    }

}
